### PR TITLE
[KIWI-1397]Create hmrcToken lambda

### PR DESCRIPTION
**Pull request by hemavana**
### What changed
Hmrc token lambda to generate access_token and store the same to the SSM.
Cron scheduled to run it every 3 hours to replace the SSM parameter token so that verifyAccount lambda can use the same.

### Why did it change

token required to access the verifyAccount api. 

### Issue tracking

- [KIWI-1397](https://govukverify.atlassian.net/browse/KIWI-1397)

### PII logging

- [x] Verified that no PII data is being logged

### Environment variables or secrets

Added a few env variables required for this lambda.

[KIWI-1397]: https://govukverify.atlassian.net/browse/KIWI-1397?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: deploy/template.yaml [hunk 1]
```diff
@@ -1125,6 +1137,175 @@ Resources:
...
+        Sourcemap: true
+        EntryPoints:
+          - HmrcTokenHandler.ts
+  
+  HmrcTokenScheduledRule: 
+    Type: AWS::Events::Rule
+    Condition: NotLocalTestStack
+    Properties: 
+      Description: "ScheduledRule"
+      ScheduleExpression: cron(0/5 * * * ? *)
```
Comment (hemavana) [inline deploy/template.yaml:1191]: I have added the cron to run every 5 minutes for now to test main stack deployment. Once its working, will update it to cron(0 0/3 * * ? *) and remove the commented lines. Currently main stack lambda is trying to execute but failing for PutParameter due to permissionsBoundary error.

File: deploy/template.yaml [hunk 2]
```diff
@@ -1125,6 +1137,175 @@ Resources:
...
+      - "HmrcTokenLogGroup"
+    Properties:
+      FunctionName: !Sub "HmrcToken-${AWS::StackName}"
+      CodeUri: ../src/
+      Handler: HmrcTokenHandler.lambdaHandler
+      ReservedConcurrentExecutions: !If
+        - ApplyReservedConcurrency
+        - !Ref LambdaConcurrency
+        - !Ref AWS::NoValue
+      # Events:
```
Comment (ViveakV) [inline deploy/template.yaml:1153]: Guessing this is no longer needed so can we remove please?

File: deploy/template.yaml [hunk 3]
```diff
@@ -1125,6 +1137,175 @@ Resources:
...
+      ScheduleExpression: cron(0/5 * * * ? *)
+      State: "ENABLED"
+      Targets: 
+        - 
+          Arn: 
+            Fn::GetAtt: 
+              - "HmrcTokenFunction"
+              - "Arn"
+          Id: "HmrcTokenFunction"
+  # PermissionForEventsToInvokeLambda:
```
Comment (ViveakV) [inline deploy/template.yaml:1200]: Is this meant to be commented out? Looks like it's needed to actually enable the cron?

Comment (hemavana) [reply deploy/template.yaml:1200]: we have HmrcTokenFunctionPermission defined which gives permissions to all resources and this wont be required. I just commented it to test it in main stack before i confirm and remove it.

File: deploy/template.yaml (hunk 2 above)
Comment (hemavana) [reply deploy/template.yaml:1153]: Yes, plan is to remove this. I just want to get this working on main BE stack before i confirm and remove these commented unnecessary code.

Comment (hemavana) [reply deploy/template.yaml]: Removed as we have tested this in main stack and its working

File: deploy/template.yaml (hunk 3 above)
Comment (hemavana) [reply deploy/template.yaml]: Removed the commented lines.

File: deploy/template.yaml [hunk 4]
```diff
@@ -1512,6 +1530,160 @@ Resources:
...
+  
+  HmrcTokenScheduledRule: 
+    Type: AWS::Events::Rule
+    Condition: NotLocalTestStack
+    Properties: 
+      Description: "ScheduledRule"
+      ScheduleExpression: cron(0 0/3 * * ? *)
+      State: "ENABLED"
+      Targets: 
+        -
```
Comment (ViveakV) [inline deploy/template.yaml:1582]: Is this right? looked weird to me

File: src/HmrcTokenHandler.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+	private readonly HMRC_CLIENT_ID_SSM_PATH = checkEnvironmentVariable(EnvironmentVariables.HMRC_CLIENT_ID_SSM_PATH, logger);
+
+	private readonly HMRC_CLIENT_SECRET_SSM_PATH = checkEnvironmentVariable(EnvironmentVariables.HMRC_CLIENT_SECRET_SSM_PATH, logger);
+
+	async handler(): Promise<void> {
+		logger.setPersistentLogAttributes({});
+		
+		try {
+			logger.info("Generating a new HMRC token");
+			if (!HMRC_CLIENT_ID) {
```
Comment (ViveakV) [inline src/HmrcTokenHandler.ts:33]: Should these individual try catches be nested inside the parent try catch? Might end up throwing multiple errors for 1 issue cause we're catching the throw

File: src/services/HmrcTokenRequestProcessor.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+		this.logger.info("Generating a new hmrc access token ");
+		try {
+			const data = await this.hmrcService.generateToken();
+			
+			if (!data) {
+				throw new AppError(HttpCodesEnum.SERVER_ERROR, "Failed generating hmrc access token");
+			}
+
+			// Validating expires_in 
+			if (data.expires_in !== Constants.HMRC_EXPECTED_TOKEN_EXPIRES_IN) {
```
Comment (ViveakV) [inline src/services/HmrcTokenRequestProcessor.ts:45]: Nice!

File: src/HmrcTokenHandler.ts (hunk 5 above)
Comment (hemavana) [reply src/HmrcTokenHandler.ts:34]: refactored the code to abstract the ssm fetch to a new function and letting the global try/catch to catch the AppError and action it.

**Review by ViveakV — APPROVED** [review]

**Pull request merged**